### PR TITLE
Insert externals directly in the editor

### PR DESCRIPTION
**Pull request by nosolosw**
# How to test

* Open the media modal, choose Google source, and select 1 picture. Confirm button should have "Insert" text and upon clicking it, should insert the image as a transient in the editor.
* Open the media modal, choose Google source, and select 2 or more pictures. Confirm button should have "Copy to library" text and upon clicking it, should copy items to library and change the screen to the media library.
* Check that set feature image and site settings icon behavior is not affected.

Also, run automatic tests:

    npm run test-client client/post-editor/media-modal/test/index.jsx

### CODE REVIEW

**Comment by nosolosw** [comment]
Rebased from master, that already includes changes in PR https://github.com/Automattic/wp-calypso/pull/17386

File: client/post-editor/media-modal/index.jsx [hunk 1]
```diff
@@ -427,7 +443,7 @@ export class EditorMediaModal extends Component {
 		if ( this.state.source !== '' ) {
 			buttons.push( {
 				action: 'confirm',
-				label: this.props.translate( 'Copy to media library' ),
+				label: selectedItems.length > 1 ? this.props.translate( 'Copy to media library' ) : this.props.translate( 'Insert' ),
```
Comment (nosolosw) [inline client/post-editor/media-modal/index.jsx]: AFAIK, the site icon modal gets its button name [from here](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/site-settings/site-icon-setting/index.jsx#L310), and the set featured image [from here](https://github.com/Automattic/wp-calypso/blob/master/client/post-editor/editor-featured-image/index.jsx#L90) - so this change doesn't affect them.

I've also tested and it works for me. Do you get this text on the site icon and featured image modals?

**Comment by nosolosw** [comment]
Good catch on videos! I'll review that use case: besides not allowing transient videos to be inserted, we need to prevent videos from being copied should your plan not support them. Will also work on fixing 2.

Couldn't reproduce 3, though, but I get the "unable to retrieve your full library" error quite often (mostly after loading a few pages). Was this what you got? I guess that is related to the Google API malfunctioning? Can't see how that would be related to these changes, it'd be lovely if you could take a look at that.

error-unable-to-retrieve

File: client/post-editor/media-modal/index.jsx (hunk 1 above)
Comment (johngodley) [inline client/post-editor/media-modal/index.jsx]: > AFAIK, the site icon modal gets its button name from here, and the set featured image from here - so this change doesn't affect them.

Yep, but what I mean is that you need to allow that prop to override the value here which is hardcoded to either 'Copy to media library' or 'Insert'. This is used elsewhere in the same function:

```js
	label: this.props.labels.confirm || this.props.translate( 'Insert' )
```

The text for the featured image and site icon is incorrect for me when viewing Google Photos. It is correct when viewing internal

**Comment by johngodley** [comment]
> besides not allowing transient videos to be inserted, we need to prevent videos from being copied should your plan not support them

I think that should already be handled and an error will be shown.

Regarding issue 3, no that's a different error and there's nothing much we can do about that. Either way I don't think it's related to this PR, but I'll try and confirm.

Also, minor thing, but the commit message for 'Add tests' could be made a bit more descriptive!

**Comment by nosolosw** [comment]
> Also, minor thing, but the commit message for 'Add tests' could be made a bit more descriptive!

:sweat_smile:

**Comment by johngodley** [comment]
I've not been able to reproduce the exact thing I saw, even when failing an upload.

Talking of failing uploads if the upload fails on the site icon then it silently ignores it, and moves on to the edit page. If you cancel that you see the upload message, if you continue then it works anyway as the resize code will copy a new resized version into the media library.

If you insert a photo directly and it fails then no error is shown but it is immediately removed from the editor. However it is still visible in the HTML editor. I suspect this behaviour also exists if you drop an image directly into the post editor (i.e. no media library) and it fails to upload - will double check.

Note an easy way to force a failed upload is to edit `lib/media/actions.js` and change the `getExternalUploader` function to something like this (note the extra text at the end)

```js
const getExternalUploader = service => ( file, siteId ) => {
	return wpcom.undocumented().site( siteId ).uploadExternalMedia( service, [ file.guid + 'zzzz' ] );
};
```

**Comment by johngodley** [comment]
> I suspect this behaviour also exists if you drop an image directly into the post editor (i.e. no media library) and it fails to upload - will double check.

No, it removes the failed upload from both visual and text editor. To test this edit `lib/media/actions.js` again and in `uploadFiles` drop an exception in the callback:

```js
			return uploader( file, siteId ).then( ( data ) => {
				throw 'Error';
				Dispatcher.handleServerAction( Object.assign( action, {
					data: data.media[ 0 ]
				} ) );
				// also refetch media limits
				Dispatcher.handleServerAction( {
					type: 'FETCH_MEDIA_LIMITS',
					siteId: siteId
				} );
			} ).catch( ( error ) => {
				Dispatcher.handleServerAction( Object.assign( action, { error } ) );
			} );
```

If you then drop an image onto the editor it will appear and then get removed once uploaded.

**Comment by nosolosw** [comment]
* 7443a32 prevents the uploads to be shown in the google modal if you cancel the site icon process.
* 2cf6e22 fixes the labels.
* 009c8a1 changes behavior for videos: user is prompted to copy them to library instead of directly inserting them.

**Comment by nosolosw** [comment]
I'm working on incorporating the external use case for failing uploads.

This is what happens now if a local upload fails:

* inserted in the editor: it's removed (both visual and HTML).
* featured image: it's removed.
* site icon: the images appear on the editing screen, but clicking the Done button does nothing (shows no errors, do not confirm and set the image either).

The site icon behavior is a bit weird and we probably want to modify that as well for local images.

**Comment by nosolosw** [comment]
fc46213 removes the external from visual mode as well as editor mode (if the upload failed).

File: client/post-editor/media-modal/index.jsx [hunk 2]
```diff
@@ -152,9 +175,17 @@ export class EditorMediaModal extends Component {
...
 			);
-			this.copyExternal( itemsWithTransientId, this.state.source );
+			if ( itemsWithTransientId.length === 1 && MediaUtils.getMimePrefix( itemsWithTransientId[ 0 ] ) === 'image' ) {
+				this.copyExternal( itemsWithTransientId, this.state.source );
+				this.props.onClose( {
+					type: 'media',
+					items: itemsWithTransientId
+				} );
+			} else {
+				this.copyExternalAfterLoadingWordPressLibrary( itemsWithTransientId, this.state.source );
```
Comment (gwwar) [inline client/post-editor/media-modal/index.jsx]: In the future please use early returns over if/else nesting.

**Pull request merged**